fvm/evm/emulator/state/base.go

```diff
@@ -0,0 +1,472 @@
+	code, err := v.GetCode(addr)
+	return len(code), err
+}
+
+// GetState returns values for an slot in the main storage
+//
+// for non-existent slots it returns the default empty hash value (gethTypes.EmptyCodeHash)
+func (v *BaseView) GetState(sk types.SlotAddress) (gethCommon.Hash, error) {
+	return v.getSlot(sk)
+}
+
+// UpdateSlot updates the value for a slot
+func (v *BaseView) UpdateSlot(sk types.SlotAddress, value gethCommon.Hash) error {
+	return v.storeSlot(sk, value)
+}
+
+// GetRefund returns the total amount of (gas) refund
+//
+// this method returns the value of zero
+func (v *BaseView) GetRefund() uint64 {
+	return 0
+}
+
+// GetTransientState returns values for an slot transient storage
+//
+// transient storage is not a functionality for the base view so it always
+// returns the default value for non-existent slots
+func (v *BaseView) GetTransientState(types.SlotAddress) gethCommon.Hash {
+	return gethCommon.Hash{}
+}
+
+// AddressInAccessList checks if an address is in the access list
+//
+// access list control is not a functionality of the base view
+// it always returns false
+func (v *BaseView) AddressInAccessList(gethCommon.Address) bool {
+	return false
+}
+
+// SlotInAccessList checks if a slot is in the access list
+//
+// access list control is not a functionality of the base view
+// it always returns false
+func (v *BaseView) SlotInAccessList(types.SlotAddress) (addressOk bool, slotOk bool) {
+	return false, false
+}
+
+// CreateAccount creates a new account
+func (v *BaseView) CreateAccount(
+	addr gethCommon.Address,
+	balance *big.Int,
+	nonce uint64,
+	code []byte,
+	codeHash gethCommon.Hash,
+) error {
+	var colID []byte
+	// if is an smart contract account
+	if len(code) > 0 {
+		err := v.storeCode(addr, code)
+		if err != nil {
+			return err
+		}
+
+		// create a new collection for slots
+		col, err := v.collectionProvider.NewCollection()
+		if err != nil {
+			return err
+		}
+		colID = col.CollectionID()
+	}
+
+	// create a new account and store it
+	acc := NewAccount(addr, balance, nonce, codeHash, colID)
+
+	// no need to update the cache , storeAccount would update the cache
+	return v.storeAccount(acc)
+}
+
+// UpdateAccount updates an account's meta data
+func (v *BaseView) UpdateAccount(
+	addr gethCommon.Address,
+	balance *big.Int,
+	nonce uint64,
+	code []byte,
+	codeHash gethCommon.Hash,
+) error {
+	acc, err := v.getAccount(addr)
+	if err != nil {
+		return err
+	}
+	// if update is called on a non existing account
+	// we gracefully call the create account
+	// TODO: but we might need to revisit this action in the future
+	if acc == nil {
+		return v.CreateAccount(addr, balance, nonce, code, codeHash)
+	}
+	// if it has a code change
+	if codeHash != acc.CodeHash {
+		err := v.storeCode(addr, code)
+		if err != nil {
+			return err
+		}
+		// TODO: maybe purge the state in the future as well
+		// currently the behaviour of stateDB doesn't purge the data
+	}
```

Comment:
Should we remove slots collection if `code` is empty?

Reply:
that shouldn't happen, if the code is empty it should not be any slots,

Reply:
> that shouldn't happen, if the code is empty it should not be any slots,

Is it possible for `UpdateAccount()` to be called with empty `code` after account was created with non-empty `code`?  If so, does `UpdateAccount()` need to remove the allocated slots collection?

Reply:
I don't think it is possible right now, even updating the account with an updated code for an account is not an option for the EVM. I added some details in the doc.